read-fonts/src/lib.rs

```diff
@@ -71,6 +71,86 @@ pub(crate) mod codegen_prelude {
 
 include!("../generated/font.rs");
 
+#[derive(Clone)]
+/// Reference to the content of a font or font collection file.
+pub enum FileRef<'a> {
+    /// A single font.
+    Font(FontRef<'a>),
+    /// A collection of fonts.
+    Collection(CollectionRef<'a>),
+}
+
+impl<'a> FileRef<'a> {
+    pub fn new(data: FontData<'a>) -> Result<Self, ReadError> {
+        Ok(if let Ok(collection) = CollectionRef::new(data) {
+            Self::Collection(collection)
+        } else {
+            Self::Font(FontRef::new(data)?)
+        })
+    }
+
+    pub fn fonts(&self) -> impl Iterator<Item = Result<FontRef<'a>, ReadError>> + 'a + Clone {
+        let copy = self.clone();
+        let len = match self {
+            Self::Collection(collection) => collection.len(),
+            _ => 1,
+        };
+        (0..len).filter_map(move |ix| match &copy {
+            Self::Font(font) => Some(Ok(font.clone())),
+            Self::Collection(collection) => collection.get(ix),
+        })
+    }
+}
+
+const TTC_HEADER_TAG: Tag = Tag::new(b"ttcf");
+
+/// Reference to the content of a font collection file.
+#[derive(Clone)]
+pub struct CollectionRef<'a> {
+    data: FontData<'a>,
+    header: TTCHeader<'a>,
+}
+
+impl<'a> CollectionRef<'a> {
+    pub fn new(data: FontData<'a>) -> Result<Self, ReadError> {
+        let header = TTCHeader::read(data)?;
+        if header.ttc_tag() != TTC_HEADER_TAG {
+            Err(ReadError::InvalidTtc(header.ttc_tag()))
+        } else {
+            Ok(Self { data, header })
+        }
+    }
+
+    pub fn len(&self) -> u32 {
+        self.header.num_fonts()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn get(&self, index: u32) -> Option<Result<FontRef<'a>, ReadError>> {
```

Comment:
I could maybe make an argument for not doing the `Option<Result<_>>` dance here, and just say that passing an invalid index is a valid error condition (in a way that it is not an error condition of a nullable offset is null).

Reply:
This seems reasonable to me